rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,227 @@
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels: # List of labels to select desired Kubernetes services
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`kubernetes_cluster` field in the config will be translated into `teleport.dev/kubernetes-cluster` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+#### Annotations
+
+Kubernetes annotations will allow users to fine tune transformation of services into Teleport apps. They will override default behaviour, 
+but they are not required for import of services - by default services without any annotations will also be imported.
+
+Annotation `teleport.dev/discovery-type` controls what Teleport entity will be created from this service, for now we will only support value 
+`app` which means Teleport application will be imported from this service. In the future there are plans to expand to database importing.
+If annotation is not set service is treated as an app.
+
+Annotation `teleport.dev/discovery-protocol` controls protocol for the access of the Teleport app we create. If annotation is missing
+default fallback for the the app's URI is `tcp`. However, if annotation is not set, additional attempts will be performed to try to 
+determine protocol of an exposed port:
+-  if kubernetes service port definition has `appProtocol` field, and it contains
+values `http`/`https` we will use it in the URI.
+- if exposed port's name is `http` or it has numeric value 80, `http` will be used.
+- if exposed port's name is `https` or it has numeric value 443, `https` will be used.
+
+Annotation `teleport.dev/discovery-port` controls preferred port for the Kubernetes service, only this one will be used even if service
+has multiple exposed ports. Its value should be one of the exposed service ports; otherwise, the app will not be imported. 
+Value can be matched either by numeric value or by the name of the port defined on the service.
+
+Annotation `teleport.dev/app-rewrite` controls rewrite configuration for Teleport app, if needed. It should
```

Comment:
I'd say initially yes, and later we can iterate on complicating annotations capabilities. My thinking is that it shouldn't be too bad, since probably majority of apps don't require two http type ports with different rewrite rules, what do you think?

Reply:
Looks good, thank you :pray:

The rewrite rules are, IMO, not very scalable from UX PoV :thinking: 
I think we could come up with a newer resource that would store RewriteRuleTemplates
This way we could apply the same recipe to multiple apps of the same family.
Eg, every Grafana will require the same set of rules

Out of scope for the current RFD, but we might think about this in the future.